src/utils/display-observer.js

```diff
@@ -0,0 +1,248 @@
+      const newDisplay = computeDisplay(observations, this.isDocVisible_);
+      if (newDisplay != null && newDisplay !== oldDisplay) {
+        for (let k = 0; k < callbacks.length; k++) {
+          callCallbackNoInline(callbacks[k], newDisplay);
+        }
+      }
+    });
+  }
+
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @param {number} observer
+   * @private
+   */
+  observed_(entries, observer) {
+    const seen = new Set();
+    for (let i = entries.length - 1; i >= 0; i--) {
+      const {target, isIntersecting} = entries[i];
+      if (seen.has(target)) {
+        continue;
+      }
+      seen.add(target);
+      const callbacks = this.targetObserverCallbacks_.get(target);
+      if (!callbacks) {
+        continue;
+      }
+      let observations = this.targetObservations_.get(target);
+      if (!observations) {
+        observations = [];
+        this.targetObservations_.set(target, observations);
+      }
+      const oldDisplay = computeDisplay(observations, this.isDocVisible_);
+      observations[observer] = isIntersecting;
+      const newDisplay = computeDisplay(observations, this.isDocVisible_);
+      if (newDisplay != null && newDisplay !== oldDisplay) {
+        for (let k = 0; k < callbacks.length; k++) {
+          callCallbackNoInline(callbacks[k], newDisplay);
+        }
+      }
+    }
+  }
+}
+
+/**
+ * @param {!Element} target
+ * @param {function(boolean)} callback
+ */
+export function observeDisplay(target, callback) {
+  getObserver(target).observe(target, callback);
+}
+
+/**
+ * @param {!Element} target
+ * @param {!ObserverCallbackDef} callback
+ */
+export function unobserveDisplay(target, callback) {
+  getObserver(target).unobserve(target, callback);
+}
+
+/**
+ * @param {!Element} target
+ * @return {!Promise<boolean>}
+ */
+export function measureDisplay(target) {
+  const observer = getObserver(target);
+  return new Promise((resolve) => {
+    const onDisplay = (display) => {
+      resolve(display);
+      observer.unobserve(target, onDisplay);
+    };
+    observer.observe(target, onDisplay);
+  });
+}
+
+/**
+ * @param {!Element} target
+ * @return {!DisplayObserver}
+ */
+function getObserver(target) {
+  registerServiceBuilderForDoc(target, SERVICE_ID, DisplayObserver);
+  return /** @type {!DisplayObserver} */ (getServiceForDoc(target, SERVICE_ID));
+}
+
+/**
+ * @param {!ObserverCallbackDef} callback
+ * @param {../layout-rect.LayoutSizeDef} size
+ */
+function callCallbackNoInline(callback, size) {
+  try {
+    callback(size);
+  } catch (e) {
+    rethrowAsync(e);
+  }
+}
+
+/**
+ * @param {?Array<boolean>} observations
+ * @param {boolean} isDocVisible
+ * @return {?boolean}
+ */
+function computeDisplay(observations, isDocVisible) {
+  if (!isDocVisible) {
+    return false;
+  }
+  if (!observations) {
+    // Unknown yet.
+    return null;
+  }
+  if (observations[DOC_OBSERVER] || observations[VIEWPORT_OBSERVER]) {
```

Comment:
Continuing https://github.com/ampproject/amphtml/pull/31976#discussion_r560639551:

I still don't like this tuple treatment. I think if we're hardcoding `DOC_OBSERVER` and `VIEWPORT_OBSERVER` as indexes, we'd be better served by using a struct with keys. If we want to make this a generic "array of observers and array of their observations", we should remove the hardcoded indexes and use generic `push`/`forEach`/`reduce` to add/loop/compute.

Reply:
Ok. Switch it fully to array/loop/reduce now.